### PR TITLE
Feature/dynamic price

### PR DESCRIPTION
**Pull request by malteish**
see #199 
## Changes
An interface is defined that will provide the fundraising contract with additional price information. The fundraising contract enforces it's own bounds on the price, to prevent unforeseen problems.

`PriceLinearTime` is included as a demonstration of how such a price oracle could work.

If a price oracle wants to use information like `tokensSold` for it's calculations, it can do so by reading this from the calling contract.

## Open questions
`PriceLinearTime` is pretty large at the moment, as it is ownable and supports ERC2771. This is only required to make it reusable. We could also decide to use each oracle only once, which would make it much smaller. Especially since the parameters are stored in one slot, this approach is worth a thought.

### CODE REVIEW

File: contracts/PriceLinearTime.sol [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+ *      The currency the offer is denominated in is set at creation time and can be updated later.
+ *      The contract can be paused at any time by the owner, which will prevent any new deals from being made. Then, changes to the contract can be made, like changing the currency, price or requirements.
+ *      The contract can be unpaused after "delay", which will allow new deals to be made again.
+ *      A company will create only one PublicFundraising contract for their token (or one for each currency if they want to accept multiple currencies).
+ * @dev The contract inherits from ERC2771Context in order to be usable with Gas Station Network (GSN) https://docs.opengsn.org/faq/troubleshooting.html#my-contract-is-using-openzeppelin-how-do-i-add-gsn-support
+ */
+contract PriceLinearTime is
+    ERC2771ContextUpgradeable,
+    Ownable2StepUpgradeable,
+    PausableUpgradeable,
```
Comment (CJentzsch) [inline contracts/PriceLinearTime.sol:31]: Why does it need to be pausable, and what does this even mean in this context?

Comment (malteish) [reply contracts/PriceLinearTime.sol]: You are very right, this is totally not necessary.
Removed it: 80eb5082becc0b8de5129f1ece6c0db8795c85b8

File: contracts/PriceLinearTime.sol [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+ *      The currency the offer is denominated in is set at creation time and can be updated later.
+ *      The contract can be paused at any time by the owner, which will prevent any new deals from being made. Then, changes to the contract can be made, like changing the currency, price or requirements.
+ *      The contract can be unpaused after "delay", which will allow new deals to be made again.
+ *      A company will create only one PublicFundraising contract for their token (or one for each currency if they want to accept multiple currencies).
+ * @dev The contract inherits from ERC2771Context in order to be usable with Gas Station Network (GSN) https://docs.opengsn.org/faq/troubleshooting.html#my-contract-is-using-openzeppelin-how-do-i-add-gsn-support
+ */
+contract PriceLinearTime is
+    ERC2771ContextUpgradeable,
+    Ownable2StepUpgradeable,
+    ReentrancyGuardUpgradeable,
```
Comment (CJentzsch) [inline contracts/PriceLinearTime.sol:30]: Nutzen wir den ReentrancyGuard irgendwo hier?

Comment (CJentzsch) [reply contracts/PriceLinearTime.sol:30]: Oder heißt das wenn man davon erbt wir er automatisch in jeder Funktion genutzt?

Comment (malteish) [reply contracts/PriceLinearTime.sol:30]: Der ... wird auch nicht gebraucht. :/

Comment (malteish) [reply contracts/PriceLinearTime.sol]: Auch unnötig, wurde entfernt. Danke für den Hinweis!

**Review by CJentzsch — DISMISSED** [review]

File: contracts/PriceLinear.sol [hunk 3]
```diff
@@ -0,0 +1,154 @@
...
+    uint64 slopeEnumerator;
+    /// denominator of slope of linear function, , e.g. b where slope == a/b
+    uint64 slopeDenominator;
+    /// start time as unix timestamp or start block number
+    uint64 start;
+    /// step width in seconds or blocks
+    uint32 stepDuration;
+    /// if 0: `stepDuration` is in seconds and `start` is an epoch
+    /// if 1: `stepDuration` is in blocks and `start` is a block number
+    uint8 isBlockBased;
```
[the remark this answers is not in the document]
Comment (malteish) [reply contracts/PriceLinear.sol]: absolutely, changed it. [9bfc903](https://github.com/corpus-io/tokenize.it-smart-contracts/pull/210/commits/9bfc90374e579ada8b1dce5701d9d9695b2c1728)

File: contracts/PriceLinear.sol [hunk 4]
```diff
@@ -0,0 +1,154 @@
...
+    /// if 0 price is falling, if 1 price is rising
+    uint8 isRising;
```
[the remark this answers is not in the document]
Comment (malteish) [reply contracts/PriceLinear.sol]: absolutely, changed it. [9bfc903](https://github.com/corpus-io/tokenize.it-smart-contracts/pull/210/commits/9bfc90374e579ada8b1dce5701d9d9695b2c1728)

**Review by CJentzsch — APPROVED** [review]

**Pull request merged**